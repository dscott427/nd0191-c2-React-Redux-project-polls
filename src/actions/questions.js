import { saveQuestion, saveQuestionAnswer } from "../utils/api";
import { showLoading, hideLoading } from "react-redux-loading-bar";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ADD_QUESTION = "ADD_QUESTION";
export const ADD_QUESTION_ANSWER = "ADD_QUESTION_ANSWER";

function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

export function handleAddQuestion(question) {
  return async (dispatch, getState) => {
    const { authedUser } = getState();

    dispatch(showLoading());

    const question_1 = await saveQuestion({
      question,
      author: authedUser,
    });
    dispatch(addQuestion(question_1));
    return dispatch(hideLoading());
  };
}

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

function addQuestionAnswer({ id, authedUser, answer }) {
  return {
    type: ADD_QUESTION_ANSWER,
    id,
    authedUser,
    answer,
  };
}

export function handleAddQuestionAnswer(id, answer, getState) {
  return (dispatch) => {
    const { authedUser } = getState();
    dispatch(addQuestionAnswer(id, authedUser, answer));

  /*   return saveLikeToggle(info).catch((e) => {
      console.warn("Error in handleToggleTweet: ", e);
      dispatch(toggleTweet(info));
      alert("The was an error liking the tweet. Try again.");
    }); */
  };
}
