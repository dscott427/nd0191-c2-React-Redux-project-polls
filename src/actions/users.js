export const RECEIVE_USERS = "RECEIVE_USERS";
export const ADD_USER_ANSWER = "ADD_USER_ANSWER";
export const ADD_USER_POLL = "ADD_USER_POLL";


export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function addUserPoll(users) {
  return {
    type: ADD_USER_POLL,
    users,
  };
}

export function addUserAnswer({ id, authedUser, answer }) {
  return {
    type: ADD_USER_ANSWER,
    id,
    authedUser,
    answer,
  };
}

export function handleAddUserAnswer(question) {

  console.log("handleAddUserAnswer Action");
  return async (dispatch,getState) => {
    const { authedUser } = getState();

    let userAnswer = {
      id: question.id,
      authedUser: question.authedUser,
      answer: question.option,
    }

    dispatch(addUserAnswer(userAnswer));
  }
}
