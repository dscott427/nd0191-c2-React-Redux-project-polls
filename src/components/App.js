import { useEffect, Fragment } from "react";
import { connect } from "react-redux";
import { handleInitialData } from "../actions/shared";
import Dashboard from "./Dashboard";
import LoadingBar from "react-redux-loading-bar";
import Question from "./Question";
import QuestionCreate from "./QuestionCreate";
import Leaderboard from "./Leaderboard";
import Login from "./Login";
import Logout from "./Logout";
import Nav from "./Nav";
import { Routes, Route } from "react-router-dom";

const App = (props) => {
  useEffect(() => {
    props.dispatch(handleInitialData());
  }, []);

  return (
    <Fragment>
      <LoadingBar />
      <div className="container">
        <Nav />
        {props.loading === true ? null : (
          <Routes>
            <Route path="/" exact element={<Dashboard />} />
            <Route path="/question/new" element={<QuestionCreate />} />
            <Route path="/question/answer" element={<Question />} />
            <Route path="/leaderboard" element={<Leaderboard />} />
            <Route path="/login" element={<Login />} />
            <Route path="/logout" element={<Logout />} />
          </Routes>
        )}
      </div>
    </Fragment>
  );
};

const mapStateToProps = ({ authedUser }) => ({
  loading: authedUser === null,
});

export default connect(mapStateToProps)(App);
