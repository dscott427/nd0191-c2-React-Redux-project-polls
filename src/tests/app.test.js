import React from "react";
import { render } from "@testing-library/react";
import App from "../components/App";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { setAuthedUser } from "../actions/authedUser";
import '@testing-library/jest-dom/extend-expect';
import { createStore } from "redux";
import reducer from "../reducers";
import middleware from "../middleware";
import { handleInitialData } from "../actions/shared";

const path = "/";
const match = {
    isExact: true,
    path,
    url: path
};

const store = createStore(reducer, middleware);



describe('App', () => {

    it("should render the App", () => {

        const view = render(
            <Provider store={store}>
                <BrowserRouter>
                    <App />
                </BrowserRouter>
            </Provider>
        );
        expect(view).toBeDefined();
        expect(view).toMatchSnapshot();
    });

    it("should render login", () => {
        const view = render(
          <Provider store={store}>
            <BrowserRouter>
              <App />
            </BrowserRouter>
          </Provider>
        );
        
        expect(view.queryByTestId("login-component")).toBeInTheDocument();

      });

      it("should render dashboard on successful logon", async () => {
        await store.dispatch(setAuthedUser("tylermcginnis"));
        await store.dispatch(handleInitialData());
   
        const view = render(
          <Provider store={store}>
            <BrowserRouter>
              <App />
            </BrowserRouter>
          </Provider>
        );
    
        const heading = view.getByTestId("dashboard-component");
        expect(heading).toBeInTheDocument();
      });

});