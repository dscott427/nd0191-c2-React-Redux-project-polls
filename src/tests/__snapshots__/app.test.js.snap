// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render the App 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="direction: ltr;"
      >
        <div
          class=""
          style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
        />
        <div
          style="display: table; clear: both;"
        />
      </div>
      <div
        class="container"
      >
        <nav
          class="nav"
        >
          <ul>
            <li>
              <a
                href="/"
                style="pointer-events: none;"
              >
                Home
              </a>
            </li>
            <li>
              <a
                href="/leaderboard"
                style="pointer-events: none;"
              >
                Leaderboard
              </a>
            </li>
            <li>
              <a
                href="/add"
                style="pointer-events: none;"
              >
                New
              </a>
            </li>
            <li>
              <a
                href="/"
              >
                Login
              </a>
            </li>
            <li>
              <img
                alt=""
                class="avatarsmall"
                src=""
              />
            </li>
          </ul>
        </nav>
        <div
          data-testid="login-component"
        >
          <h3
            class="center"
          >
            Log In
          </h3>
          <form>
            <div>
              User
            </div>
            <input
              class="text"
              data-testid="login-user-name"
              id="optionOne"
              maxlength="280"
              placeholder="User"
              type="text"
              value=""
            />
            <div>
              Password
            </div>
            <input
              class="text"
              data-testid="login-password"
              id="optionTwo"
              maxlength="280"
              placeholder="Password"
              type="text"
              value=""
            />
            <div>
              <button
                class="btn"
                data-testid="login-submit-button"
                disabled=""
                type="submit"
              >
                Submit
              </button>
            </div>
            <div />
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="direction: ltr;"
    >
      <div
        class=""
        style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
      />
      <div
        style="display: table; clear: both;"
      />
    </div>
    <div
      class="container"
    >
      <nav
        class="nav"
      >
        <ul>
          <li>
            <a
              href="/"
              style="pointer-events: none;"
            >
              Home
            </a>
          </li>
          <li>
            <a
              href="/leaderboard"
              style="pointer-events: none;"
            >
              Leaderboard
            </a>
          </li>
          <li>
            <a
              href="/add"
              style="pointer-events: none;"
            >
              New
            </a>
          </li>
          <li>
            <a
              href="/"
            >
              Login
            </a>
          </li>
          <li>
            <img
              alt=""
              class="avatarsmall"
              src=""
            />
          </li>
        </ul>
      </nav>
      <div
        data-testid="login-component"
      >
        <h3
          class="center"
        >
          Log In
        </h3>
        <form>
          <div>
            User
          </div>
          <input
            class="text"
            data-testid="login-user-name"
            id="optionOne"
            maxlength="280"
            placeholder="User"
            type="text"
            value=""
          />
          <div>
            Password
          </div>
          <input
            class="text"
            data-testid="login-password"
            id="optionTwo"
            maxlength="280"
            placeholder="Password"
            type="text"
            value=""
          />
          <div>
            <button
              class="btn"
              data-testid="login-submit-button"
              disabled=""
              type="submit"
            >
              Submit
            </button>
          </div>
          <div />
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
